### PR TITLE
Desktop: Fixes #4891: Solve "Resource Id not provided" error

### LINKED ISSUE #4891: Resource ID not provided error when click link with markdown pane closed
A markdown link in one folder that points to a note in another folder does not work and throws a "Resource ID not provided" error when the markdown pane is closed (only the WYSIWYG editor is visible) but works fine when the markdown pane is visible.

## Environment

Joplin version: 
Platform: 
OS specifics: Windows 10 64-bit

Joplin 1.7.11 (prod, win32)

Client ID: 447e55cf95a4441ca8a84eb7ace2bb05
Sync Version: 2
Profile Version: 34
Keychain Supported: Yes

Revision: f560563d8 (master)

## Steps to reproduce

1. Right-click on an existing note
2. Copy markdown link
3. Move to a new notebook
4. Create a new note
5. Paste the link into the markdown editor -- e.g., [Regression](:/99675cd0851a41af807c00390a27d30f)
6. At this point the markdown link appears in the Markdown editor and a clickable link labeled Regression appears in the WYSIWYG editor with the Joplin icon to the left of the link.

image

7. As long as the editor is toggled so that the markdown editor is in the left pane and the WYSIWYG editor is in the right pane, clicking on the link in the WYSIWYG pane behaves normally--clicking on the link takes you to the Joplin note. But...
8. Toggle the editor off (so that only the WYSIWYG editor is visible and the markdown editor is hidden.
9. Click the link in the WYSIWYG editor and nothing happens.
10. Right click and choose Open and an error window appears "Resource ID not provided"

image
 

## Describe what you expected to happen

I expect that clicking on the WYSIWYG link would take me to the Joplin note regardless of whether the markdown editor pane is open or closed.

### PR DESCRIPTION
**Pull request by Subhra264**
Fixes: #4891 

A function `openItemById` is created inside `app-desktop/gui/NoteEditor/utils/contextMenu.ts` just to avoid duplicate code. The `useMessageHandler` hook and  `menuItems`  both use the `openItemById` function so that the user gets the same experience on `ctrl+click` or `right-click+open` a link.

# Testing
### Manual Testing
Manually tested both on markdown editor and the TinyMCE editor( with note-links, resource-links; performing both the operations: `ctrl+click` and ` right-click+open` ).

### Unit testing
Sorry ( again ), I couldn't find a way to write a unit test for it. Maybe it is possible to add a test, but I think it would be very ( very) complex ( so need some guide ).

### CODE REVIEW

File: packages/app-desktop/gui/NoteEditor/utils/contextMenu.ts [hunk 1]
```diff
@@ -53,17 +55,52 @@ function handleCopyToClipboard(options: ContextMenuOptions) {
 	}
 }
 
-export function menuItems(): ContextMenuItems {
+export async function openItemById(options: any, dispatch: Function) {
```
Comment (laurent22) [inline packages/app-desktop/gui/NoteEditor/utils/contextMenu.ts:58]: Looking at the code, you only need an ID and an optional hash, so pass this to the function explicitely: `itemId: string, dispatch:Function, hash:string = ''` Please do not use "any" as it disables type checking.

File: packages/app-desktop/gui/NoteEditor/utils/contextMenu.ts [hunk 2]
```diff
@@ -53,17 +55,52 @@ function handleCopyToClipboard(options: ContextMenuOptions) {
...
+	const itemId = options.resourceId || options.itemId;
+	const item = await BaseItem.loadItemById(itemId);
+
+	if (!item) throw new Error(`No item with ID ${itemId}`);
+
+	if (item.type_ === BaseModel.TYPE_RESOURCE) {
+		const localState = await Resource.localState(item);
```
Comment (laurent22) [inline packages/app-desktop/gui/NoteEditor/utils/contextMenu.ts:65]: Declare the resource here, with type checking: `const resource = item as ResourceEntity` and use that resource variable for this block of code.

File: packages/app-desktop/gui/NoteEditor/utils/contextMenu.ts [hunk 3]
```diff
@@ -53,17 +55,52 @@ function handleCopyToClipboard(options: ContextMenuOptions) {
...
+		}
+
+		try {
+			await ResourceEditWatcher.instance().openAndWatch(item.id);
+		} catch (error) {
+			console.error(error);
+			bridge().showErrorMessageBox(error.message);
+		}
+	} else if (item.type_ === BaseModel.TYPE_NOTE) {
+		let noteItem: any = {
```
Comment (laurent22) [inline packages/app-desktop/gui/NoteEditor/utils/contextMenu.ts:82]: It's not a "noteItem", it's an action so please call it that.

File: packages/app-desktop/gui/NoteEditor/utils/contextMenu.ts [hunk 4]
```diff
@@ -53,17 +55,52 @@ function handleCopyToClipboard(options: ContextMenuOptions) {
...
+			type: 'FOLDER_AND_NOTE_SELECT',
+			folderId: item.parent_id,
+			noteId: item.id,
+		};
+
+		if (options.hash) {
+			noteItem = { ...noteItem, hash: options.hash };
+		}
```
Comment (laurent22) [inline packages/app-desktop/gui/NoteEditor/utils/contextMenu.ts:90]: Make no sense to recreate the object. Just pass the hash directly to the action, as in the original code.

File: packages/app-desktop/gui/NoteEditor/utils/contextMenu.ts (hunk 3 above)
Comment (laurent22) [reply packages/app-desktop/gui/NoteEditor/utils/contextMenu.ts:82]: And type checking on the note: `const note = item as NoteEntity`

File: packages/app-desktop/gui/NoteEditor/utils/contextMenu.ts (hunk 2 above)
Comment (Subhra264) [reply packages/app-desktop/gui/NoteEditor/utils/contextMenu.ts]: Done

File: packages/app-desktop/gui/NoteEditor/utils/contextMenu.ts (hunk 3 above)
Comment (Subhra264) [reply packages/app-desktop/gui/NoteEditor/utils/contextMenu.ts]: > And type checking on the note `const note = item as NoteEntity`

Done

**Comment by laurent22** [comment]
All good now, thank you for the update!

**Pull request merged**